File: 2025/JuyoungMoon/week10/2025-08-10-house-robber.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 10: DP
// https://github.com/juyomo/leetcode-study

// LeetCode #198.
// https://leetcode.com/problems/house-robber/

class Solution {
public:
    int findMax(int a, int b) {
        return a > b ? a : b;
    }
    int rob(vector<int>& nums) {
        int size = nums.size();
        if (size == 0) {
            return 0;
        } else if (size == 1) {
            return nums[0];
        }

        vector<int> max(size);
        max[0] = nums[0];
        max[1] = findMax(nums[0], nums[1]);

        for(int i = 2; i < size; i++) {
            max[i] = findMax(nums[i] + max[i-2], max[i-1]);
        }

        return max[size-1];
    }
};

File: 2025/jimmycho/week10/2025-08-10-House-Robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) < 2:
            return nums[0]
        temp = [0] * (len(nums) + 1)
        temp[1], temp[2] = nums[0], nums[1]
        for i in range(3, len(nums) + 1):
            temp[i] = max(nums[i - 1] + temp[i - 2], nums[i - 1] + temp[i - 3])
        return max(temp[-2], temp[-1])